zhaquirks/tuya/ts1201.py

```diff
@@ -0,0 +1,524 @@
+https://github.com/Koenkk/zigbee-herdsman-converters/blob/9d5e7b902479582581615cbfac3148d66d4c675c/lib/zosung.js
+"""
+import base64
+from typing import Any, List, Optional, Union
+
+from zigpy.profiles import zha
+from zigpy.quirks import CustomCluster, CustomDevice
+import zigpy.types as t
+from zigpy.zcl import foundation
+from zigpy.zcl.clusters.general import (
+    Basic,
+    GreenPowerProxy,
+    Groups,
+    Identify,
+    OnOff,
+    Ota,
+    PowerConfiguration,
+    Scenes,
+    Time,
+)
+
+from zhaquirks.const import (
+    DEVICE_TYPE,
+    ENDPOINTS,
+    INPUT_CLUSTERS,
+    MODELS_INFO,
+    OUTPUT_CLUSTERS,
+    PROFILE_ID,
+)
+
+
+class Bytes(bytes):
+    """Bytes serializable class."""
+
+    def serialize(self):
+        """Serialize Bytes."""
+        return self
+
+    @classmethod
+    def deserialize(cls, data):
+        """Deserialize Bytes."""
+        return cls(data), b""
+
+
+class ZosungIRControl(CustomCluster):
+    """Zosung IR Control Cluster (0xE004)."""
+
+    name = "Zosung IR Control Cluster"
+    cluster_id = 0xE004
+    ep_attribute = "zosung_ircontrol"
+
+    attributes = {0x0000: ("last_learned_ir_code", t.CharacterString, True)}
+
+    server_commands = {
+        0x00: foundation.ZCLCommandDef(
+            "data",
+            schema={"data": Bytes},
+            direction=foundation.Direction.Server_to_Client,
+            is_manufacturer_specific=True,
+        ),
+        0x01: foundation.ZCLCommandDef(
+            "IRLearn",
+            schema={"on_off": t.Bool},
+            direction=foundation.Direction.Server_to_Client,
+            is_manufacturer_specific=True,
+        ),
+        0x02: foundation.ZCLCommandDef(
+            "IRSend",
+            schema={"code": t.CharacterString},
+            direction=foundation.Direction.Server_to_Client,
+            is_manufacturer_specific=True,
+        ),
+    }
+
+    def _update_attribute(self, attrid, value):
+        super()._update_attribute(attrid, value)
+
+    async def read_attributes(
+        self, attributes, allow_cache=False, only_cache=False, manufacturer=None
+    ):
+        """Read attributes ZCL foundation command."""
+        if 0x0000 in attributes:
+            return (
+                {0: self.endpoint.device.last_learned_ir_code},
+                {},
+            )
+        else:
+            attr = await super().read_attributes(
+                attributes,
+                allow_cache=allow_cache,
+                only_cache=only_cache,
+                manufacturer=manufacturer,
+            )
+            return attr
+
+    async def command(
+        self,
+        command_id: Union[foundation.GeneralCommand, int, t.uint8_t],
+        *args,
+        manufacturer: Optional[Union[int, t.uint16_t]] = None,
+        expect_reply: bool = True,
+        tsn: Optional[Union[int, t.uint8_t]] = None,
+        **kwargs: Any,
+    ):
+        """Override the default Cluster command."""
+        if command_id == 1:
```

Comment:
The [new `ServerCommandDefs`](https://github.com/zigpy/zha-device-handlers/pull/2336/files#r1375463975) could be used here (`name.id`) instead of repeating the command id again here.